crates/fast_compute/src/layer.rs

```diff
@@ -622,11 +619,68 @@ impl<'a, T: TowerFamily, P: PackedTop<T>> ComputeLayerExecutor<T::B128>
 
 	fn pairwise_product_reduce(
 		&mut self,
-		_input: <Self::DevMem as ComputeMemory<T::B128>>::FSlice<'_>,
-		_round_outputs: &mut [<Self::DevMem as ComputeMemory<T::B128>>::FSliceMut<'_>],
+		input: <Self::DevMem as ComputeMemory<T::B128>>::FSlice<'_>,
+		round_outputs: &mut [<Self::DevMem as ComputeMemory<T::B128>>::FSliceMut<'_>],
 	) -> Result<(), Error> {
-		// TODO(CRY-490)
-		todo!()
+		let log_num_inputs = match strict_log_2(input.len()) {
+			None => {
+				return Err(Error::InputValidation(format!(
+					"input length must be a power of 2: {}",
+					input.len()
+				)));
+			}
+			Some(0) => {
+				return Err(Error::InputValidation(format!(
+					"input length must be greater than or equal to 2 in order to perform at least one reduction: {}",
+					input.len()
+				)));
+			}
+			Some(log_num_inputs) => log_num_inputs,
+		};
+		let expected_round_outputs_len = log_num_inputs;
+		if round_outputs.len() != expected_round_outputs_len as usize {
```

Comment:
nit: maybe move the logic verifying the input and output dimensions to a separate helper function to share between the reference and fast implementations?